yarn-project/slasher/src/slasher_client.ts

```diff
@@ -0,0 +1,377 @@
+          this.addMonitoredPayload(payload).catch(e => {
+            this.log.error('Error adding monitored payload', e);
+          });
+        }
+        this.sortMonitoredPayloads();
+      },
+    });
+  }
+
+  /**
+   * Convert a list of factory events to an iterable of monitored payloads
+   *
+   * @param args
+   * @returns the list of monitored payloads
+   */
+  private *factoryEventsToMonitoredPayloads(
+    args: GetContractEventsReturnType<typeof SlashFactoryAbi, 'SlashPayloadCreated'>,
+  ): IterableIterator<MonitoredSlashPayload> {
+    for (const event of args) {
+      if (!event.args) {
+        continue;
+      }
+      const args = event.args;
+      if (!args.payloadAddress || !args.validators || !args.amounts || !args.offences) {
+        continue;
+      }
+
+      yield {
+        payloadAddress: EthAddress.fromString(args.payloadAddress),
+        validators: args.validators.map(EthAddress.fromString),
+        amounts: args.amounts,
+        offenses: args.offences.map(offense => bigIntToOffence(offense)),
+        observedAtSeconds: this.dateProvider.now() / 1000,
+        totalAmount: args.amounts.reduce((acc, amount) => acc + amount, BigInt(0)),
+      };
+    }
+  }
+
+  /**
+   * Add a payload to the list of monitored payloads if we agree with it
+   *
+   * @param payload
+   */
+  private async addMonitoredPayload(payload: MonitoredSlashPayload) {
+    if (await this.doIAgreeWithPayload(payload)) {
+      this.log.info('Adding monitored payload', payload);
+      this.monitoredPayloads.push(payload);
+    } else {
+      this.log.info('Disagreeing with payload', payload);
+    }
+  }
+
+  /**
+   * Check if we agree with a payload
+   *
+   * We check each offense and validator pair against the watchers
+   *
+   * @param payload
+   * @returns true if any watcher agrees with the payload, false otherwise
+   */
+  private async doIAgreeWithPayload(payload: MonitoredSlashPayload) {
+    // zip offenses and validators together
+    const offensesAndValidators = payload.offenses.map((offense, index) => ({
+      offense,
+      validator: payload.validators[index],
+      amount: payload.amounts[index],
+    }));
+
+    // check each offense
+    for (const offenseAndValidator of offensesAndValidators) {
+      const watcherResponses = await Promise.all(
+        this.watchers.map(watcher =>
+          watcher.shouldSlash(
+            offenseAndValidator.validator.toString(),
+            offenseAndValidator.amount,
+            offenseAndValidator.offense,
+          ),
+        ),
+      );
+      // if no watcher agrees, return false
+      if (watcherResponses.every(response => !response)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Sort the monitored payloads by total amount in descending order
+   */
+  private sortMonitoredPayloads() {
+    this.monitoredPayloads.sort((a, b) => Number(b.totalAmount) - Number(a.totalAmount));
```

Comment:
You gotta use a bigint down here, otherwise you will run into a world of pain as soon as you try a read slashing amount, e.g. remember that 1 token is 1e18 which is too large to fit into a number safely (2**53 – 1).